Lint with flake8

feerrenrut opened the pull request:
<!--
Please fill in the following template, for an explanation of the sections see:
https://github.com/nvaccess/nvda/wiki/Github-pull-request-template-explanation-and-examples
-->

### Link to issue number:
Closes #5918

### Summary of the issue:
Code contributors regularly have to deal with ill-defined and inconsistently enforced code style requirements. Code reviewers spend much of their time reporting minor issues, time that would be better spent looking for architectural problems / product issues / logic errors.

### Description of how this pull request fixes the issue:
Automate checking python style. The diff from new PR's will be tested for compliance with Flake8. The NVDA Python code already contains several inconsistent styles, so rather than try to match it I have tried to configure Flake8 to use the default style guidelines as much as possible.

This pr introduces two new SCons build targets:
- `lint`
  - creates a unified diff with `git diff -U0 $(git merge-base <baseBranch>)`
    - A helper script is used to generate this diff (`tests\lint\genDiff.py`)
  - The diff is piped to `flake8` to perform the linting.
  - The output is printed to stdout and also to `tests/lint/current.lint`
- `lintInstall`
  - required by `lint`.
  - Uses pip to install dependencies from a `requirements.txt` file.

AppVeyor:
- Adds a new script for tests phase of build
- Mostly does what SCons does, does not need to worry about getting working tree / uncommit changes into the diff.
- In order to preserve the availability of artifacts, these are uploaded from a `on_finish` phase rather than `artifacts` phase.
  - This acts like a "finally" block, and happens regardless of whether the build passes or fails.  
  - The installer artifact is often used to test if a change fixes an issue before the PR is polished off / reviewed. It also can help reviewers to test a change locally without having to build the branch.
- A message is sent when there are linting errors.
- A failed lint will still halt the build, system tests are not run.

### Testing performed:
Tested locally:
- Enured that uncommited and unstaged changes in the working tree are included in diff given to Flake8
- Checked that `#noqa: <code>` comments work to suppress errors.

Tested on appveyor:
- Sample output for build with errors: https://github.com/nvaccess/nvda/pull/9958#issuecomment-516534268
- Sample output for clean build: Search for `flake8.lint.flake8_diff_lint` on https://ci.appveyor.com/project/NVAccess/nvda/builds/26350748/tests

### Known issues with pull request:
- [x] Add documentation to readme.
- [x] Test file to demonstrate, needs removal.
- [x] Using `--disable-noqa` to demonstrate test file, needs removal
- Alters the users installed packages, see `lintInstall`
- Requires `git` in path.

### Change log entry:
Changes for developers:
` - Flake8 linting tool has been integrated with SCons reflecting code requirements for Pull Requests. (#5918)`

feerrenrut committed:
Lint with flake8 on a local diff

Uses flake8-tabs so that we can use tab rather than spaces.
Includes a default config file for flake8 (flake8.ini)

Scons build targets added:

lint:
	Produces a unified diff (against HEAD), and runs flake8 linter against this diff.
lintInstall:
	Uses pip (and a requirements file) to install flake8 dependency.

A file test.py is included as a demonstration, this will be removed prior to merging.

feerrenrut committed:
Appveyor to run lint

Test result must be converted to junit compatible xml.

feerrenrut committed:
Actually do the diff

feerrenrut committed:
Make sure we fetch the base branch first

We do a shallow clone when initially fetching the repo, so we need to get the base branch as well now.

feerrenrut committed:
Hopefully get the encoding right this time.

leonardder commented on tests/lint/createJunitReport.py:
```diff
@@ -0,0 +1,59 @@
+		out.write(outContents)
+
+
+try:
```
If you're not aware, make sure this is only run `if __name__ == "__main__":`

leonardder commented on tests/lint/flake8.ini:
```diff
@@ -0,0 +1,32 @@
+	E126, # continuation line over-indented for hanging indent
+	E133, # closing bracket is missing indentation
+
+builtins = # inform flake8 about functions we consider built-in.
```
You might have to add pgettext in here as well.

leonardder commented on tests/lint/sconscript:
```diff
@@ -0,0 +1,45 @@
+diffTarget = externalEnv.Command(
+	"current.diff",
+	None,
+	[["git", "diff", "-u", "HEAD", ">", '$TARGET']]
```
I wonder whether it would be possible to make a test against the base branch. A quick Google search turns out that this is not very easy. May be hardcode this to master of upstream (not to local master). See https://stackoverflow.com/questions/29810331/is-there-a-quick-way-to-git-diff-from-the-point-or-branch-origin

```suggestion
	[["git", "diff", "-u", "origin/master...HEAD", ">", '$TARGET']]
```

feerrenrut replied:
Yeah I might have a play with some of these options. This was really only provided as a convenience/example for developers. Note that appveyor knows the base branch and can do the diff more directly. 

Perhaps the clearest way is if we require the "upstream" to be set and generate the diff against that: `git diff @{u}...HEAD`. We can give an error message if the upstream is not set.

leonardder replied:
> Perhaps the clearest way is if we require the "upstream" to be set and generate the diff against that: `git diff @{u}...HEAD`. We can give an error message if the upstream is not set.

Mm, but if I create a topic branch and push that, setting my private fork as the upstream for that particular branch which I usually do, then `git diff @{u}...HEAD` will likely compare against upstream of the currently checked out branch. This is also not what we want I assume?

feerrenrut replied:
Yeah, you are right, every time you push this will be resetting the diff, not what we want. Instead, I think I will just require a 'ref' (branch / commit / tag) as an argument to scons, and we will do it against that branch.

leonardder replied:
Awesome idea

LeonarddeR reviewed (commented):
I'm aware of this still being a draft, but I'm pretty interested to see this finished.

feerrenrut committed:
Fix some issues with script and config

leonardder commented:
I have the following suggestions for the configuration of flake8:

* Add --count parameter to see number of errors
* add miscDeps and source\louis to exclude
* Add doctests = True

feerrenrut commented:
> Add --count parameter to see number of errors

`count` is not so helpful, we already print statistics (giving the number of each error that occurred), count prints an extra number. See the last "1" in the following example output:

```
tests/lint/createJunitReport.py:44:1: E302 expected 2 blank lines, found 1
def main():
^
1     E302 expected 2 blank lines, found 1
1
```

leonardder commented:
[**@feerrenrut**](https://github.com/feerrenrut) commented on [24 Jul 2019, 19:31 CEST](https://github.com/nvaccess/nvda/pull/9958#issuecomment-514726595 "2019-07-24T17:31:23Z - Replied by Github Reply Comments"):
> > Add --count parameter to see number of errors
> 
> `count` is not so helpful, we already print statistics (giving the number of each error that occurred), count prints an extra number. See the last "1" in the following example output:
> 
> ```
> tests/lint/createJunitReport.py:44:1: E302 expected 2 blank lines, found 1
> def main():
> ^
> 1     E302 expected 2 blank lines, found 1
> 1
> ```

Ah, I had hoped that it was more descriptive. Never mind.

feerrenrut committed:
Allow specifying base ref for scons based lint

feerrenrut committed:
Fix malformed command line for flake8

feerrenrut committed:
When lint is not provided, we still get an error about the missing base argument.

There might be a better way to do this. It would be nice if arguments
were tied to sub scripts.

leonardder commented on appveyor.yml:
```diff
@@ -157,6 +160,25 @@ test_script:
+      git diff -u FETCH_HEAD...HEAD > $prDiff
+      $flake8Config = "$lintSource\flake8.ini"
+      $flake8Output = "$lintOutput\PR-Flake8.txt"
+      type "$prDiff" | py -m flake8 --exit-zero --disable-noqa --diff --output-file="$flake8Output" --tee --config="$flake8Config"
```
Is your usage of --exit-zero intentional?

feerrenrut replied:
Yes, for now, I'm testing it out. It would be nice if lint errors did not stop the build from completing, so we get access to build artifacts etc. But we still need to be notified of the failures.

feerrenrut replied:
First I want to confirm that this is the thing stopping the build with a mysterious "Command exited with code 1" after the system tests are done.

leonardder commented on appveyor.yml:
```diff
@@ -157,6 +160,25 @@ test_script:
+      $lintSource = (Resolve-Path .\tests\lint\)
+      git fetch -q origin $env:APPVEYOR_REPO_BRANCH
+      $prDiff = "$lintOutput\prDiff.patch"
+      git diff -u FETCH_HEAD...HEAD > $prDiff
```
The problem with the current diff approach is that flake8 also parses the untouched lines in the diff and prints warnings about them. I think this might be fine at some point, but if we ever decide that linting must be ok before we allow a pull request to pass, it requires the author to touch lots of code that is part of the context rather than the actually changed code. And.. you know what's going to happen, people update the context, then new context is added to the diff which also results into warnings.

This can be fixed by getting less lines of context in the diff, e.g.

```suggestion
      git diff -U0 -u FETCH_HEAD...HEAD > $prDiff
```

This is pretty extreme though.

```suggestion
      git diff -U1 -u FETCH_HEAD...HEAD > $prDiff
```

This might be a nice intermediate choice.

feerrenrut replied:
Good point, I think U0 is the better choice. Start with the most minimal requirement. The intention is that linting should pass before the final review / merge. We may not be strict about this to start with, but its the goal.

feerrenrut committed:
Use zero context in unified diff

Errors should not be raised for adjacent code.

feerrenrut committed:
Try to ensure that build artifacts are available even on lint failure.

Also send a message when linting fails.

feerrenrut committed:
Include stats in lint message

feerrenrut committed:
Leave stats out for now.

feerrenrut committed:
Clarify flake8 error message

leonardder commented on tests/lint/sconscript:
```diff
@@ -0,0 +1,61 @@
+baseBranch = env.get("base")
+if doLint and not baseBranch:
+	errorMessage = (
+		"Lint can not complete without base branch."
```
```suggestion
		"Lint can not complete without base branch. "
```

leonardder commented on tests/lint/sconscript:
```diff
@@ -0,0 +1,61 @@
+if doLint and not baseBranch:
+	errorMessage = (
+		"Lint can not complete without base branch."
+		"Try: 'scons lint base=origin/master'"
```
```suggestion
		"Try: 'scons lint base=origin/master'. "
```

feerrenrut committed:
Spacing in missing base argument error message

Co-Authored-By: Leonard de Ruijter <leonardder@users.noreply.github.com>

feerrenrut committed:
Spacing in missing base argument error message

Co-Authored-By: Leonard de Ruijter <leonardder@users.noreply.github.com>

leonardder commented:
Testing with flake8 within VS Code reveals that it raises warnings for using `_` and `pgettext`. I'm afraid this will also be the case for appveyor builds. It is likelike that scons has sourceEnv imported before it uses flake8

feerrenrut commented:
Are you using the config file from this PR?

leonardder commented:
Ah I'm sorr, turns out that I was using an out of date version of the branch for my testing purposes.

feerrenrut committed:
update readme to include new scons targets

feerrenrut committed:
Generate diff including working tree

feerrenrut committed:
Remove example

Remove --disable-noqa
Remove test.py full of # noqa: comments

feerrenrut committed:
Ignore Flake8 warning: W503

W503 is: line break before binary operator.
Contrasted with W504: line break after binary operator, which we want to check for.

leonardder commented on appveyor.yml:
```diff
@@ -180,9 +206,9 @@ deploy_script:
+      $apiVersion = (py -c "import sys; sys.path.append('source'); from addonAPIVersion import CURRENT; print('{}.{}.{}'.format(*CURRENT))")
       echo apiversion: $apiVersion
-      $apiCompatTo = (python -c "import sys; sys.path.append('source'); from addonAPIVersion import BACK_COMPAT_TO; print('{}.{}.{}'.format(*BACK_COMPAT_TO))")
+      $apiCompatTo = (py -c "import sys; sys.path.append('source'); from addonAPIVersion import BACK_COMPAT_TO; print('{}.{}.{}'.format(*BACK_COMPAT_TO))")
```
It would be awesome if we could get these two versions in one go instead of calling the python script twice. Not that I want this for this pr, though.

feerrenrut replied:
We could return them as a string and process it... but that has it's own downsides.

leonardder commented on tests/lint/flake8.ini:
```diff
@@ -0,0 +1,35 @@
+exclude = # don't bother looking in the following subdirectories / files.
+	.git,
+	__pycache__,
+	.tox,
```
Do we have this dir?

feerrenrut replied:
Possibly not, but since there is no point looking in these directories why would we let it?

leonardder commented on tests/lint/genDiff.py:
```diff
@@ -0,0 +1,75 @@
+]
+
+
+def substitutePlaceholder(
```
Why don't you use python string formatting here? e.g.:

```
_mergeBaseCommand = "git merge-base {baseBranchPlaceholder,  # this is the target branch used in a PR} HEAD"
```

then, you could call format on `_mergeBaseCommand` to pass it the actual branch you want to fill in.
Or, is it really required to use a list here?

Honestly, I really feel a bit uncomfortable with looking up the placeHolderIndex like this.

feerrenrut replied:
I did it this way to match the examples for `subprocess.check_output`. I also don't really like to have to look up the index, I wanted the commands at the beginning of the file to make it clear what the point was. I'll see if using string.format works.

feerrenrut replied:
I decided to keep the args in list, but join them with space (so we don't have to remember to put a space after each argument manually), then format the string.

leonardder commented on tests/lint/genDiff.py:
```diff
@@ -0,0 +1,75 @@
+def getDiff(baseBranch: str) -> bytes:
+	mergeBaseCommand = substitutePlaceholder(_mergeBaseCommand, _baseBranchPlaceholder, baseBranch)
+	mergeBase: bytes = subprocess.check_output(mergeBaseCommand)
+	mergeBase: str = mergeBase.decode(encoding='ANSI').strip()
```
Are you sure ANSI is the right encoding here? SHouldn't this be `locale.getpreferredencoding()` or locale.getlocale()[1]? Note that the former properly takes the Windows 10 beta unicode support into account.

feerrenrut replied:
The output of this is a SHA and is to be consumed by an application. I think variations based on lang/locale would only be likely to cause problems.

leonardder commented on tests/lint/lintInstall/requirements.txt:
```diff
@@ -0,0 +1,5 @@
+####### requirements.txt #######
+#
+###### Requirements for automated lint ######
+flake8 == 3.7.7
```
3.7.8 is now out. May be go for the most recent version while at it. I understand that we couldn't keep on changing the version.

See also https://pip.pypa.io/en/stable/reference/pip_install/#id29

```suggestion
flake8 ~= 3.7.7
```

This ensures a compatible release is installed.

feerrenrut replied:
> This ensures a compatible release is installed.

The intention is a compatible release, it relies on the judgement of the human setting the version number. That said, a breakage here is likely to be fairly low impact for us, so I think this makes sense.

leonardder commented on tests/lint/sconscript:
```diff
@@ -0,0 +1,60 @@
+	raise SCons.SConf.SConfError(errorMessage)
+
+# Create a unified diff. Written to file for ease of manual inspection
+diffTarget = externalEnv.Command(
```
Have you considered just importing the genDiff module and executing the diff creation from there? We use python calls python several times in our scons environment and I think it is ok, but it has its downsides.

feerrenrut replied:
This needs to be run in the external environment, so that it has access to the path for git.

leonardder commented on tests/lint/sconscript:
```diff
@@ -0,0 +1,60 @@
+		'--output-file=$TARGET',  # output to a file to allow easier inspection
+		'--tee',  # also output to stdout, so results appear in scons output
+		'--config="tests/lint/flake8.ini"',  # use config file of complex options
+		'--exit-zero',  # even if there are errors, don't stop the build.
```
Could this be a left over from your testing, or is this intentional? I think it makes sense tha scons fails if linting fails.

feerrenrut replied:
This was intentional, when trying to build for testing purposes it's pretty annoying if something like lint output gets in the way. Since it doesn't run by default, it's most likely that someone is running it explicitly to check the results. For now I would like to ease people into it. Though if people are finding that they not noticing the results, and we think failing the build would make it more obvious, then we can change it in the future.

LeonarddeR reviewed (commented):
I think it would be good to remove source/pylintrc while at it. According to @bramd, it has its issues anyway.

feerrenrut committed:
Update Flake8 version

Allow automatic changes to minor versions.

Co-Authored-By: Leonard de Ruijter <leonardder@users.noreply.github.com>

feerrenrut commented on tests/lint/createJunitReport.py:
```diff
@@ -0,0 +1,64 @@
+			)
+		makeJunitXML(flake8OutputFileName, junitFileName)
+	except Exception as e:
+		print(e)
```
Honestly I cant remember if this was strictly necessary. It is only used by Appveyor builds at the moment, and I'd prefer the error to show up twice than not at all.

feerrenrut committed:
Review actions

leonardder commented:
Note that the base of this is still threshold.

feerrenrut commented:
> Note that the base of this is still threshold.

Thanks 👍

josephsl commented:
Hi,

Coming back to this, I think it would be helpful to exclude checks for visual indents; although useful, it has drawbacks:

1. Some text editors may not expose indentation formatting correctly, leading some to believe the code has been indented when it didn't.
2. It produces E101 (mixture of tabs and spaces). For long lines, it can cause line length issue to pop up.
3. For long conditional statements, splitting it into multiple lines may introduce more errors than intended, and without wrapping them inside parentheses, syntax error (E999) is raised.

I'll create a new issue to discuss these further.

Thanks.

feerrenrut commented:
@josephsl we have a few issues to work through with the indentation checks, but I would really like it if we can find an acceptable configuration. This will likely require people to think about the configuration of their tools.

josephsl commented:
Hi, right. Personally I can deal with indents quite easily thanks to Notepad++, but I imagine others may run into issues (older versions of Notepad, for example). Thanks.

 

From: Reef Turner <notifications@github.com> 
Sent: Friday, August 2, 2019 9:52 AM
To: nvaccess/nvda <nvda@noreply.github.com>
Cc: Joseph Lee <joseph.lee22590@gmail.com>; Mention <mention@noreply.github.com>
Subject: Re: [nvaccess/nvda] Lint with flake8 (#9958)

 

@josephsl <https://github.com/josephsl>  we have a few issues to work through with the indentation checks, but I would really like it if we can find an acceptable configuration. This will likely require people to think about the configuration of their tools.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub <https://github.com/nvaccess/nvda/pull/9958?email_source=notifications&email_token=AB4AXEATXI3P2W45SU2JOHTQCRQZ7A5CNFSM4IF4FM72YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD3OJDFY#issuecomment-517771671> , or mute the thread <https://github.com/notifications/unsubscribe-auth/AB4AXEGVDGFGQ5PTMPU3NTTQCRQZ7ANCNFSM4IF4FM7Q> .

michaelDCurran commented:
This pr breaks publishing of NVDA releases on our server. I was able to work around this for beta1, but  this certainly needs to be addressed for the final release.
In appveyor.yaml, commands in deploy_script make use of the $artifacts variable, which is set by appveyor based on pushed artifacts. However, pushing of  artifacts in the output directory, including the NVDA launcher executable, had been moved to the on_finish section, which is run after deploy_script, thus $artifacts does not contain them. Uploading of the output directory really needs to be done probably at the end of build_script.
